discord_oauth.go

```diff
@@ -0,0 +1,178 @@
+package main
+
+import (
+	"bytes"
+	"crypto/rand"
+	"database/sql"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"io"
+	"log"
+	"net/http"
+	"net/url"
+	"time"
+)
+
+func (r *Router) DiscordSignin(w http.ResponseWriter, req *http.Request) {
+	state := generateStateOauthCookie(w)
+	redirectUri := fmt.Sprintf("%s/discord/callback", r.rootURL)
+	url := fmt.Sprintf("https://discord.com/oauth2/authorize?client_id=%s&response_type=code&redirect_uri=%v&scope=identify+guilds.join&state=%v", r.bot.ClientId, url.QueryEscape(redirectUri), state)
+	http.Redirect(w, req, url, http.StatusTemporaryRedirect)
+}
+
+func (r *Router) DiscordCallback(w http.ResponseWriter, req *http.Request) {
+	stateCookie, err := req.Cookie("oauthstate")
+	if err != nil {
+		http.Error(w, "Missing oauthstate cookie", http.StatusBadRequest)
```

Comment:
Also log when there is an http error with some context, so we can debug if there is an issue. This applies to other `http.Error`s here too